Android gradle builds

danzeeeman opened the pull request:
This adds support to build for Android using the Android Studio and the Gradle-Experimental plugin.  It is based on the work @HalfdanJ did with the gradle build process in #5059 but starting with a fresh copy of the master branch.  The PR also changes how data is handled in Android Studio and inside the app closing #2995.   The PR adds support for creating and updating projects with the project generator.  The PR makes android build faster the VS2015 builds 👍.

This also depends on an apothecary update to build projects with opencv.  I'll make that PR shortly to apothecary but I have the [libs in a zip](https://www.dropbox.com/s/kma2md0v5dbkqot/libs.zip?dl=1) for people to test apps in [my dropbox](https://www.dropbox.com/s/kma2md0v5dbkqot/libs.zip?dl=1).

This should remove a bunch of the weird android setup procedures with Eclipse and allow Android builds to be more easily maintained when the NDK is updated.  

cc: @arturoc

arturoc commented:
That sounds great, there seems to be much more than support for the experimental gradle plugin going on though. Can you detail what else is changing?

Also ping @tallavi who has done all the refactoring of the life cycle since this seems to be touching that

danzeeeman commented:
The major changes are in the way it builds.  It completely removes the old make system.   It still requires the r12 NDK. 

In the java code I added a getContext call to return the current OFActivity/Context.  I made the Activity in OFAndroidLifeCycle an OFActivity.  Also in OFAndroidLifeCycle I swap out unpacking the zip & copying data for the new copy data functions in.  It also migrates from the old data path to the new data path if the app has already been installed.

Those are the major Java side changes. 

Master had a bug where the ofAppAndroidWindow couldn't get the ofActivity as it has had moved during the refactor from OFAndroid to the OFAndroidLifeCycle object and had been renamed so I updated that in the JNI cpp.

HalfdanJ commented:
This looks good to me! 
Two comments, in my PR i had also updated the docs: https://github.com/openframeworks/openFrameworks/pull/5059/files#diff-e2236aa37a4cd501b95151cef48d5a2b and and cleaned all the examples to have the new project files.

tallavi commented:
Nice to see action in this issue! I'm here if there are any questions on the changes I did back then in the android life cycle. 

Is this PR complementary to @HalfdanJ 's one or should replace it? I tried to make the other one work on my machine (OSX) a while back and couldn't, I have an open question there regarding project generator. I see this PR targets VS, so it's not related to making it work on OSX?

danzeeeman commented:
@HalfdanJ yeah I updated the docs in the ofSite but haven't in core.  I need to cull the older project files but that is easy to fix.

danzeeeman commented:
@tallavi it is to replace @HalfdanJ's PR

apitaru commented:
Hi @danzeeeman, tried running the project on a vanilla machine but unfortunately wasn't able to compile a project as of yet using the branch.

In addition to the un-updated examples issue mentioned above, I've noticed that the branch is missing the project generator. I exported a new project using the generator in the latest osx release (0.9.8) and I got the error below:

/Users/p/dev/OF/openFrameworks-android-gradle-builds/libs/openFrameworks/utils/ofConstants.h
Error:(72, 10) 'glm/glm.hpp' file not found
Error:Execution failed for task ':openFrameworksProject:compileOpenFrameworksLibArmeabi-v7aDebugArm7StaticLibraryOpenFrameworksLibArmeabi-v7aDebugArm7StaticLibraryMainCpp'.
> Multiple build operations failed.
      C++ compiler failed while compiling of3dUtils.cpp.
      C++ compiler failed while compiling ofMainLoop.cpp.
      C++ compiler failed while compiling ofCamera.cpp.
      C++ compiler failed while compiling ofNode.cpp.
      C++ compiler failed while compiling ofEasyCam.cpp.
      C++ compiler failed while compiling of3dPrimitives.cpp.
      C++ compiler failed while compiling ofAppNoWindow.cpp.
      C++ compiler failed while compiling ofAppRunner.cpp.
      C++ compiler failed while compiling ofSerial.cpp.
      C++ compiler failed while compiling ofArduino.cpp.
      ...and 48 more failures.
  See the complete log at: file:///Users/p/dev/OF/openFrameworks-android-gradle-builds/libs/openFrameworksCompiled/project/android/build/tmp/compileOpenFrameworksLibArmeabi-v7aDebugArm7StaticLibraryOpenFrameworksLibArmeabi-v7aDebugArm7StaticLibraryMainCpp/output.txt

danzeeeman commented:
this issue has been addressed.

On Tue, Mar 7, 2017 at 11:43 AM, Amit Pitaru <notifications@github.com>
wrote:

> Hi @danzeeeman <https://github.com/danzeeeman>, tried running the project
> on a vanilla machine but unfortunately wasn't able to compile a project as
> of yet using the branch.
>
> In addition to the un-updated examples issue mentioned above, I've noticed
> that the branch is missing the project generator. I exported a new project
> using the generator in the latest osx release (0.9.8) and I got the error
> below:
>
> /Users/p/dev/OF/openFrameworks-android-gradle-builds/libs/openFrameworks/
> utils/ofConstants.h
> Error:(72, 10) 'glm/glm.hpp' file not found
> Error:Execution failed for task ':openFrameworksProject:
> compileOpenFrameworksLibArmeabi-v7aDebugArm7StaticLibraryOpenF
> rameworksLibArmeabi-v7aDebugArm7StaticLibraryMainCpp'.
>
> Multiple build operations failed.
> C++ compiler failed while compiling of3dUtils.cpp.
> C++ compiler failed while compiling ofMainLoop.cpp.
> C++ compiler failed while compiling ofCamera.cpp.
> C++ compiler failed while compiling ofNode.cpp.
> C++ compiler failed while compiling ofEasyCam.cpp.
> C++ compiler failed while compiling of3dPrimitives.cpp.
> C++ compiler failed while compiling ofAppNoWindow.cpp.
> C++ compiler failed while compiling ofAppRunner.cpp.
> C++ compiler failed while compiling ofSerial.cpp.
> C++ compiler failed while compiling ofArduino.cpp.
> ...and 48 more failures.
> See the complete log at: file:///Users/p/dev/OF/
> openFrameworks-android-gradle-builds/libs/openFrameworksCompiled/
> project/android/build/tmp/compileOpenFrameworksLibArmeabi-
> v7aDebugArm7StaticLibraryOpenFrameworksLibArmeabi-
> v7aDebugArm7StaticLibraryMainCpp/output.txt
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-284779979>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zPScb_2I9jhlzrr8zdA9oYYax-Iqks5rjYkjgaJpZM4MLZS0>
> .
>



-- 
Dan Moore
makeitdoathing.com
@theDANtheMAN
dan@makeitdoathing.com

HalfdanJ commented:
I think it would be great to keep the AndroidManifest files in the android examples to keep the permissions required

nanu-c commented:
I tried to get this working but I get 
`Gradle sync failed: Plugin with id 'com.android.model.library' not found.`
I use gradle-2.2.1 and android-ndk-r10e

danzeeeman commented:
Use android ndk r12b which is the ndk for master.
On Fri, Mar 24, 2017 at 7:53 AM Aaron <notifications@github.com> wrote:

> I tried to get this working but I get
> Gradle sync failed: Plugin with id 'com.android.model.library' not found.
> I use gradle-2.2.1 and android-ndk-r10e
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-289003705>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zN8lwfPmnM0_U6APCMCqg_fdykpeks5ro67FgaJpZM4MLZS0>
> .
>

danzeeeman commented:
Also the how to has been updated in the ofSite repo where the documentation
lives.

Dan
On Fri, Mar 24, 2017 at 11:22 AM Dan Moore <dan@makeitdoathing.com> wrote:

> Use android ndk r12b which is the ndk for master.
> On Fri, Mar 24, 2017 at 7:53 AM Aaron <notifications@github.com> wrote:
>
> I tried to get this working but I get
> Gradle sync failed: Plugin with id 'com.android.model.library' not found.
> I use gradle-2.2.1 and android-ndk-r10e
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-289003705>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zN8lwfPmnM0_U6APCMCqg_fdykpeks5ro67FgaJpZM4MLZS0>
> .
>
>

danzeeeman commented:
this should fix any outstanding issues with Manifests and Documentation.  I pulled what was in the ofSite update into /docs and added back in the manifests for example that need more permissions than the base permissions basically: Audio Example, Camera Examples, and Vibrator Example.  

I'm going to merge to master unless the core group has issues cc:@openframeworks/android

HalfdanJ commented:
Great! I'll try and run the examples tomorrow

nanu-c commented:
Oh yes, the examples are working. This my first of I got working on android! Juhuhuhu

HalfdanJ commented:
@danzeeeman what about the travis check, have you had any luck getting that to work? Seems like its failing

danzeeeman commented:
@HalfdanJ I don't know enough about the travis setup to comment on the failures.

https://travis-ci.org/openframeworks/openFrameworks/branches

HalfdanJ commented:
https://travis-ci.org/openframeworks/openFrameworks/jobs/214706587

It tries to build the empty example: 

```
cd $ROOT/examples/android/androidEmptyExample
echo "ABIS_TO_COMPILE_DEBUG = $1" >> config.make
make Debug PLATFORM_OS=Android
make: *** No rule to make target `Debug'.  Stop.
```

arturoc commented:
it seems you've removed the makefiles which the test servers relied on to test some examples. i think it would be a good idea to leave the makefiles in for a couple of versions but in any case if gradle is now the main building system the test scripts should be migrated to use that before merging this or even add a new target that tests using gradle and leave the old ones in until we completely remove makefiles

HalfdanJ commented:
Another question, do you have a PR ready for apothecary?

HalfdanJ commented:
@arturoc I'm trying to build a travis script that runs the PG on the android examples, but I have trouble getting the PG to compile on the linux box. Could we maybe make a release of projectGenerator on ci.openframeworks.cc like for vs and osx? Its a pretty big build to make just in order to generate the projects.

arturoc commented:
@HalfdanJ we don't do so until now cause the binaries in linux don't work between versions so we just compile the PG whenever someone installs OF. but if it's useful for this feel free to send a PR and make sure the vm in travis is using the same ubuntu version to compile the PG and run the android tests

tallavi commented:
Hi,

I got around to trying it. Works great on an empty project. After that, I tried a project where I included ofxPoco, but THEN I got the weird compilation errors:

/android-ndk-r12b/sources/cxx-stl/llvm-libc++/libcxx/include/cstring:77:9: error: no member named 'strcmp' in the global namespace

/../../../addons/ofxPoco/libs/poco/include/Poco/Ascii.h:137:22: error: unknown type name 'UInt32'

/Development/Android/android-ndk-r12b/sources/cxx-stl/../android/support/include/wchar.h:128:10: note: 'wmemcpy' declared here

And so on and so on.

I made sure I'm using ndk 12b. BTW, I checked the ofSite in your fork, to see if I missed anything, and it says 10e there..

Any idea what's wrong..?

Thanks,

Tal

danzeeeman commented:
I updated the docs here you must be looking someplace else.

https://github.com/danzeeeman/ofSite/blob/master/content/setup/android-studio.md

and here

https://github.com/danzeeeman/openFrameworks/blob/android-gradle-builds/docs/android_studio.md



On Thu, Mar 30, 2017 at 3:36 PM, tallavi <notifications@github.com> wrote:

> Hi,
>
> I got around to trying it. Works great on an empty project. After that, I
> tried a project where I included ofxPoco, but THEN I got the weird
> compilation errors:
>
> /android-ndk-r12b/sources/cxx-stl/llvm-libc++/libcxx/include/cstring:77:9:
> error: no member named 'strcmp' in the global namespace
>
> /../../../addons/ofxPoco/libs/poco/include/Poco/Ascii.h:137:22: error:
> unknown type name 'UInt32'
>
> /Development/Android/android-ndk-r12b/sources/cxx-stl/../
> android/support/include/wchar.h:128:10: note: 'wmemcpy' declared here
>
> And so on and so on.
>
> I made sure I'm using ndk 12b. BTW, I checked the ofSite in your fork, to
> see if I missed anything, and it says 10e there..
>
> Any idea what's wrong..?
>
> Thanks,
>
> Tal
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-290521278>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zKocfkZRYUYm83S0s6hRkKNhmBR2ks5rrARFgaJpZM4MLZS0>
> .
>



-- 
Dan Moore
makeitdoathing.com
@theDANtheMAN
dan@makeitdoathing.com

danzeeeman commented:
Poco library probably needs to be rebuild? IDK
I thought we were going away from Poco?
I don't use it anymore.

On Thu, Mar 30, 2017 at 3:36 PM, tallavi <notifications@github.com> wrote:

> Hi,
>
> I got around to trying it. Works great on an empty project. After that, I
> tried a project where I included ofxPoco, but THEN I got the weird
> compilation errors:
>
> /android-ndk-r12b/sources/cxx-stl/llvm-libc++/libcxx/include/cstring:77:9:
> error: no member named 'strcmp' in the global namespace
>
> /../../../addons/ofxPoco/libs/poco/include/Poco/Ascii.h:137:22: error:
> unknown type name 'UInt32'
>
> /Development/Android/android-ndk-r12b/sources/cxx-stl/../
> android/support/include/wchar.h:128:10: note: 'wmemcpy' declared here
>
> And so on and so on.
>
> I made sure I'm using ndk 12b. BTW, I checked the ofSite in your fork, to
> see if I missed anything, and it says 10e there..
>
> Any idea what's wrong..?
>
> Thanks,
>
> Tal
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-290521278>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zKocfkZRYUYm83S0s6hRkKNhmBR2ks5rrARFgaJpZM4MLZS0>
> .
>



-- 
Dan Moore
makeitdoathing.com
@theDANtheMAN
dan@makeitdoathing.com

tallavi commented:
@danzeeeman it's says 10e in the ofSite link. The other one says 12b.

Regarding poco, it's not part of the core, but it is a addon that gets bundled with the core..

I don't think it's a rebuild issue, it MIGHT be related to something written/not written in the addon_config.mk of the addon, probably a missing compilation flag or something..? But I have no idea what.. Help...

danzeeeman commented:
can you please show me where is says 10e?  I dont see it

danzeeeman commented:
nevermind I see it didn't get changed in trouble shooting.

danzeeeman commented:
no openCV had to get rebuild to work with this so I'm guessing libpoco does as well.

danzeeeman commented:
but I'm still under the impress that we are getting away from using Poco that's why we removed it from the core and made it an addon for legacy support of other addons that used Poco.  I don't have a problem saying.... Support for Poco has been deprecated from ofxAndroid.  

What are you using Poco on Android for?

tallavi commented:
Many many things.. I have a large project that uses poco tasks, poco xml, poco time stamps, poco data formatting, poco mutex, and more... I guess I AM one of those legacy projects...

danzeeeman commented:
@tallavi thanks for volunteering to look at it then!!  haha  I'm joking...
some Googling and it might just need a rebuild with with different flags or needs its own Gradle file to build the library correctly.   Poco's docs are so out of date they only say it needs to be build with NDK R6 etc

tallavi commented:
I'm afraid I have no idea where to start. Should I download latest sources from pocoproject and try to integrate them instead of the precompiled libs? Won't that just make the source files be includes in the resulting libOFAndroidApp.so? 

I understand that the right way to do it is with apothecary stuff (which I currently know nothing about)?

danzeeeman commented:
Yeah need to update the apothecary scripts. I'll take a look at it when I
have some free time.

Try out other things and report back any errors.
On Thu, Mar 30, 2017 at 4:18 PM tallavi <notifications@github.com> wrote:

> I'm afraid I have no idea where to start. Should I download latest sources
> from pocoproject and try to integrate them instead of the precompiled libs?
> Won't that just make the source files be includes in the resulting
> libOFAndroidApp.so?
>
> I understand that the right way to do it is with apothecary stuff (which I
> currently know nothing about)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-290532464>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zIKFg6t_Dx7tYNvpMCSGKlTVToozks5rrA4egaJpZM4MLZS0>
> .
>

arturoc commented:
although we are moving away from poco it's something that a lot of people use and it's an official addon. all official addons (except platform dependent stuff) should work before merging this

tallavi commented:
Do you think it would help if I'll pull the apothecary scripts and try to build locally? I see that it's just some make files. Why would it make any difference to rebuild it? Were they built with ndk10e and I need to rebuild with 12b? Or are there some changes that needs to be done?

Can you share what you had to change in order to build openSSL for this version?

arturoc commented:
no idea why these libs need to be recompiled, as you can see here: https://github.com/openframeworks/apothecary/blob/master/scripts/android/install.sh we are already using 12b

also the errors you are getting are happening in headers not in the linker so it's probably related to something else but no idea really

danzeeeman commented:
Arturo I was seeing strange linker and header errors with openCV which is
why it needed to be rebuilt. I think this uses a little bit different
toolchain vs the standalone. It's still kinda black magic to me a bit.

It might just need a gradle build file or to adjust the addon_config.mk
file.

I did some more digging and Poco supposedly supported building with and
Android.mk file so it might be easy rebuild with the gradle toolchain.

Dan
On Fri, Mar 31, 2017 at 5:07 AM arturo <notifications@github.com> wrote:

> no idea why these libs need to be recompiled, as you can see here:
> https://github.com/openframeworks/apothecary/blob/master/scripts/android/install.sh
> we are already using 12b
>
> also the errors you are getting are happening in headers not in the linker
> so it's probably related to something else but no idea really
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-290660340>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zIRCPsQ6_1-g-X5Rc6p2QhVKl01Kks5rrMJqgaJpZM4MLZS0>
> .
>

tallavi commented:
I got it running by making the following changes:

1) I saw that from the ndk's string.h it's getting into poco's string.h and that's making issues, so I renamed the file in poco to PocoString.h

2) Then I got similar errors just with ndk's time.h and poco data/time.h, so I renamed poco to PocoTime.h

3) Then I got errors in ofXmlPoco so I just removed the file.

Then it compiles and runs.

Obviously, that's not a solution.

I believe that issue is that the include folders in the addon (this on or all of them) are being added recuresively. In poco you just need to add poco/include to the list, not recursively. But I don't know where this change should be.

danzeeeman commented:
Does Poco namespace things? I don't remember?
On Sun, Apr 2, 2017 at 3:53 AM tallavi <notifications@github.com> wrote:

> I got it running by making the following changes:
>
>    1.
>
>    I saw that from the ndk's string.h it's getting into poco's string.h
>    and that's making issues, so I renamed the file in poco to PocoString.h
>    2.
>
>    Then I got similar errors just with ndk's time.h and poco data/time.h,
>    so I renamed poco to PocoTime.h
>    3.
>
>    Then I got errors in ofXmlPoco so I just removed the file.
>
> Then it compiles and runs.
>
> Obviously, that's not a solution.
>
> I believe that issue is that the include folders in the addon (this on or
> all of them) are being added recuresively. In poco you just need to add
> poco/include to the list, not recursively. But I don't know where this
> change should be.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-290970935>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zC02qm1AdR8bwiNGg7MEJtDNQMWAks5rr1PtgaJpZM4MLZS0>
> .
>

tallavi commented:
Yes it does. And all the includes in examples are in the form of

#include <Poco/SomeClass.h>

and not 

#include "SomeClass.h"

I bet that including sub folders is what causes this.

I tried to look in the gradle methods and figure out what actually builds the list of include directories, and if the sub folders are included, but couldn't find it. Also couldn't figure out how to print the list of include folders (like I was able to do with the old make system in adt).

arturoc commented:
we have an exception for poco in the makefiles because of this, most
libraries just work by including everything in the search path
recursively but poco needs only include to be in the search paths. since
we moved poco to an addon this is specified in the addon_config.mk file,
aren't you parsing those?

On 02.04.2017 15:45, tallavi wrote:
> Yes it does. And all the includes in examples are in the form of
> 
> #include <Poco/SomeClass.h>
> 
> and not
> 
> #include "SomeClass.h"
> 
> I bet that including sub folders is what causes this.
> 
> I tried to look in the gradle methods and figure out what actually
> builds the list of include directories, and if the sub folders are
> included, but couldn't find it. Also couldn't figure out how to print
> the list of include folders (like I was able to do with the old make
> system in adt).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-290987341>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC8cLWIgrCbCncl-Pjk3GTj1uS15Hn5ks5rr6ZsgaJpZM4MLZS0>.
>

tallavi commented:
This is in the addon:

ADDON_INCLUDES = libs/poco/include
ADDON_INCLUDES += src

Those are indeed parsed as far as I can see. 

I suppose that this alone doesn't make it recursive? 

If I'm not mistaken, it's the gradle methods part that go over addon directories and adds them to the include path recursively. The question is, what can be written in the addon_config to overwrite this.

danzeeeman commented:
This is parsing those. I'll look at it.
On Sun, Apr 2, 2017 at 10:42 AM arturo <notifications@github.com> wrote:

> we have an exception for poco in the makefiles because of this, most
> libraries just work by including everything in the search path
> recursively but poco needs only include to be in the search paths. since
> we moved poco to an addon this is specified in the addon_config.mk file,
> aren't you parsing those?
>
> On 02.04.2017 15:45, tallavi wrote:
> > Yes it does. And all the includes in examples are in the form of
> >
> > #include <Poco/SomeClass.h>
> >
> > and not
> >
> > #include "SomeClass.h"
> >
> > I bet that including sub folders is what causes this.
> >
> > I tried to look in the gradle methods and figure out what actually
> > builds the list of include directories, and if the sub folders are
> > included, but couldn't find it. Also couldn't figure out how to print
> > the list of include folders (like I was able to do with the old make
> > system in adt).
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-290987341
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAC8cLWIgrCbCncl-Pjk3GTj1uS15Hn5ks5rr6ZsgaJpZM4MLZS0
> >.
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-290990653>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zKP2qhGGX4IYjStg_QKPygIFx-JMks5rr7O-gaJpZM4MLZS0>
> .
>

arturoc commented:
the addon_config.mk file is correct i think the problem is in how the new build system is parsing it. the way this should work is:

- parse the filesystem recursively for sources in:
   - src
   - libs/*/src
   - libs/*/include (this is just includes but that way .h are added to the list of sources to show in the ide)
- parse the filesystem recursively for search paths in:
    - src
    - libs/*/src
    - libs/*/include
- parse the filesystem for libraries (*.a and *.so) in
    - libs/*/lib/android/{armv7, x86}

once the list of everything is parsed from the filesytem then parse the addon_config.mk file:
   - VAR = VALUE: substitutes the previous value with the new one
   - VAR += VALUE: adds to the previous value
   - VAR_EXCLUDE = VALUE: starts a list of values to remove from the VAR list 
  - VAR_EXCLUDE += VALUE: adds to the list of values to remove from VAR

that works for values added from the filesyste too so in an addon like:

```
ofxPoco
  |_libs
      |_poco
           |_include
               |_ Poco
                    |_....
```

the parser should first add everything recursively in libs/poco but then in the config file there's:

```
ADDON_INCLUDES = libs/poco/include
```

which substitutes the list parsed from the filesystem with libs/poco/include only

similarly if there was:

```
ADDON_EXCLUDE_INCLUDE += libs/poco/include/Poco
ADDON_EXCLUDE_INCLUDE = libs/poco/include/Poco/%
```

it would remove everything in the Poco folder including the Poco folder from the list of search paths

arturoc commented:
there's also a couple of special cases for legacy reasons, ofxXmlSettings for example has tinyxml source directly in libs so if there's source files in the libs folder directly those should be added to the sources list and the libs folder itself to the search paths

tallavi commented:
That's a good lead. 

I see that the gradle script indeed assigns on = and adds on +=. I'll try to put some logs tomorrow to see why it fails.

tallavi commented:
Well, the values seems to be parsed correctly:

ret = [libs/poco/include, src]

But after it gets that value from parsing, it iterates over them and expands each of them recursively, so there's the problem. I think it's a flow issue? From what I gather in what you wrote, it should recursively add folders in the file system, but if the value is overwritten in the addon_config, it should add them explicitly and not recursively?

tallavi commented:
I think the correct way to do it is to not pass a default value to parseAddonConfig, but instead check if the value returned is empty, THEN get the default values (libs, src) and include them recursively. If we got a value from parsing, just use it.

arturoc commented:
> From what I gather in what you wrote, it should recursively add folders in the file system, but if the value is overwritten in the addon_config, it should add them explicitly and not recursively?

kindof, addon_config has wildcards '%' which allow to specify recursively, in my second example:
```
ADDON_EXCLUDE_INCLUDE = libs/poco/include/Poco
ADDON_EXCLUDE_INCLUDE += libs/poco/include/Poco/%
```

everything in Poco is excluded recursively.

> I think the correct way to do it is to not pass a default value to parseAddonConfig, but instead check if the value returned is empty, THEN get the default values (libs, src) and include them recursively. If we got a value from parsing, just use it.

that could work but you have to take into account the exclusions so if there's anything excluded you also need to parse the fs anyway. the PG, makefiles, qtcreator, eclipse and visual studio plugins parse the file system always first.

my post above is almost pseudocode for how to parse this

also not related but in case it's missing too, take into acount that there's somethings in addons_config that are not in the filesystem like CFLAGS, CPPFLAGS, LDFLAGS

tallavi commented:
@danzeeeman I'm not sure I'll be able to fix everything big A said above, but I can put some improvements that should at least fix poco + possibly put some TODOs for what is left. 

Want me to do that and put a PR to your branch? Or you rather handle this yourself?

armadillu commented:
Just a note on this, got pretty far with the examples and they more or less work, but apart from ofxPOCO not compiling, I also can't get http connections to work. Something like:

```
string jsonData = ofLoadURL("http://api.github.com/users/armadillu/events").data.getText();
```

returns an empty string, and no warnings or errors in the log. I know nothing about Android, but I already added:

```<uses-permission android:name="android.permission.INTERNET"/>```

to the manifest, and opening a TCP socket directly (with ofxNetwork) works fine, so I do have internet access.

Digging inside ofLoadURL, I see that now that POCO is gone, libCurl is handling these... I tried using libcurl directly to see if I could trace the issue, or get an error at least

```
CURL *curl;
CURLcode res;
curl = curl_easy_init();
if(curl) {
	curl_easy_setopt(curl, CURLOPT_URL, "http://api.github.com/users/armadillu/events");
	curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
	res = curl_easy_perform(curl);
	if(res != CURLE_OK)
		ofLogError() << "curl_easy_perform() failed: " << string(curl_easy_strerror(res));
	curl_easy_cleanup(curl);
}
```

And I get ```curl_easy_perform() failed: Couldn't resolve host name```. A quick look in [StackOverflow](http://stackoverflow.com/questions/9168695/curl-fails-to-connect-to-the-internet-in-android) lead me to this response:

"I finally got this to work. In case anyone is interested it seems like this issue is present only when the jni shared lib is statically linked to curl and other dependents. It works fine when the shared libs are loaded dynamically."

I wonder if that's the issue here? I don't really know where to start... We are assessing the feasibility of using Android phones for a medium sized project that involves distributed interaction across several devices in a pretty short deadline --- But I get the feeling this might be too green to handle with OF+Android.

danzeeeman commented:
Question 1: Does it ofLoadURL() work with the old build process? i.e.
master without these updates?
Question 2: Does it ofLoadURL() work on Android in 0.9.8?

Side note....
Why are you using TCP calls and not like UDP for the distributed
interactions?   would think you would want the lowest latency communication
for interactivity?


Dan


On Tue, Apr 4, 2017 at 4:59 PM, Oriol Ferrer Mesià <notifications@github.com
> wrote:

> Just a note on this, got pretty far with the examples and they more or
> less work, but apart from ofxPOCO not compiling, I also can't get http
> connections to work. Something like:
>
> string jsonData = ofLoadURL("http://api.github.com/users/armadillu/events").data.getText();
>
> returns an empty string, and no warnings or errors in the log. I know
> nothing about Android, but I already added:
>
> <uses-permission android:name="android.permission.INTERNET"/>
>
> to the manifest, and opening a TCP socket directly (with ofxNetwork) works
> fine, so I do have internet access.
>
> Digging inside ofLoadURL, I see that now that POCO is gone, libCurl is
> handling these... I tried using libcurl directly to see if I could trace
> the issue, or get an error at least
>
> CURL *curl;
> CURLcode res;
> curl = curl_easy_init();
> if(curl) {
> 	curl_easy_setopt(curl, CURLOPT_URL, "http://api.github.com/users/armadillu/events");
> 	curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
> 	res = curl_easy_perform(curl);
> 	if(res != CURLE_OK)
> 		ofLogError() << "curl_easy_perform() failed: " << string(curl_easy_strerror(res));
> 	curl_easy_cleanup(curl);
> }
>
> And I get curl_easy_perform() failed: Couldn't resolve host name. A quick
> look in StackOverflow
> <http://stackoverflow.com/questions/9168695/curl-fails-to-connect-to-the-internet-in-android>
> lead me to this response:
>
> "I finally got this to work. In case anyone is interested it seems like
> this issue is present only when the jni shared lib is statically linked to
> curl and other dependents. It works fine when the shared libs are loaded
> dynamically."
>
> I wonder if that's the issue here? I don't really know where to start...
> We are assessing the feasibility of using Android phones for a medium sized
> project that involves distributed interaction across several devices in a
> pretty short deadline --- But I get the feeling this might be too green to
> handle with OF+Android.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-291630308>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zLwHCHSP7OJG_XXefJflgXh74LlGks5rsq9KgaJpZM4MLZS0>
> .
>



-- 
Dan Moore
makeitdoathing.com
@theDANtheMAN
dan@makeitdoathing.com

armadillu commented:
1: I gave up on the old process, I couldn't get the android plugin to work at all in eclipse (up-to-date master or release) following the guide currently in the OF Site. The only way I could get OF in the phone was with your new "gradle" way, so I can't answer there. 

2: There will be OSC for intra-device comm and syncing up, but we still have to download tons of assets to each device, and we'll do standard JSON parsing + http downloads of those.

danzeeeman commented:
Ahh word!


I'll look at the HTTP stuff soon.

On Tue, Apr 4, 2017 at 6:34 PM, Oriol Ferrer Mesià <notifications@github.com
> wrote:

> 1: I gave up on the old process, I couldn't get the android plugin to work
> at all in eclipse (up-to-date master or release) following the guide
> currently in the OF Site. The only way I could get OF in the phone was with
> your new "gradle" way, so I can't answer there.
>
> 2: There will be OSC for intra-device comm and syncing up, but we still
> have to download tons of assets to each device, and we'll do standard JSON
> parsing + http downloads of those.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openframeworks/openFrameworks/pull/5461#issuecomment-291656087>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAr6zLC4usAV8xRvQl676Gff-phFVlmZks5rssWSgaJpZM4MLZS0>
> .
>



-- 
Dan Moore
makeitdoathing.com
@theDANtheMAN
dan@makeitdoathing.com

danzeeeman commented:
I fixed the ofxPoco thing and pushed it.
Still haven't dug into the CURL thing but I call that a different bug not associated with the updates.

arturoc commented:
Thanks Dan, i'll look into the curl issue when i have a moment, one last thing before merging this is fixing the CI tests which are currently broken

hamoid commented:
@danzeeeman In https://github.com/danzeeeman/openFrameworks/blob/android-gradle-builds/docs/android_studio.md it says "Use project generator ..." but by default I believe the PG is not compiled. Should one compile it first? If I'm running Linux, I should run `scripts/linux/compilePG.sh`?

The first point "Install latest version of Android Studio" could give some more details, as installing Android Studio will ask about which components to add, and the user may not know what to choose.

HalfdanJ commented:
I've fixed CI, its now making the androidEmptyExample using the projectGenerator, and compiling using gradle.
I've also removed a large amount of files in the examples folder that either belong to eclipse or aren't necessary, since they are the same as the ones in the template, so will be provided when the projectGenerator is running.

danzeeeman commented:
we should merge now that its passing CI?
cc:@arturoc

arturoc commented:
that looks great thanks to everyone involved in making this happen!
